Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,78 @@
+#include "Spire/Ui/DropDownList.hpp"
+#include <QEvent>
+#include <QHBoxLayout>
+#include "Spire/Ui/ListItem.hpp"
+#include "Spire/Ui/ListView.hpp"
+#include "Spire/Ui/OverlayPanel.hpp"
+#include "Spire/Ui/ScrollableListBox.hpp"
+
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  auto SCROLLABLE_LIST_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(BackgroundColor(QColor::fromRgb(0, 0, 0, 0))).
+      set(border_size(0));
+    return style;
+  }
+
+  auto get_border_size(QWidget& widget) {
+    auto border_size = QSize(0, 0);
+    auto& stylist = find_stylist(widget);
+    auto block = stylist.get_computed_block();
+    for(auto& property : block) {
+      property.visit(
+        [&] (const BorderTopSize& size) {
+          stylist.evaluate(size, [&] (auto size) {
+            border_size.rheight() += size;
+          });
+        },
+        [&] (const BorderRightSize& size) {
+          stylist.evaluate(size, [&] (auto size) {
+            border_size.rwidth() += size;
+          });
+        },
+        [&] (const BorderBottomSize& size) {
+          stylist.evaluate(size, [&] (auto size) {
+            border_size.rheight() += size;
+          });
+        },
+        [&] (const BorderLeftSize& size) {
+          stylist.evaluate(size, [&] (auto size) {
+            border_size.rwidth() += size;
+          });
+        });
+    }
+    return border_size;
+  }
+}
+
+DropDownList::DropDownList(ListView& list_view, QWidget* parent)
+    : QWidget(parent),
+      m_list_view(&list_view) {
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  auto scrollable_list_box = new ScrollableListBox(*m_list_view, this);
+  set_style(*scrollable_list_box,
+    SCROLLABLE_LIST_STYLE(get_style(*scrollable_list_box)));
+  layout->addWidget(scrollable_list_box);
+  m_panel = new OverlayPanel(this, parent);
+  m_panel->set_closed_on_blur(true);
+  m_panel->layout()->itemAt(0)->widget()->setSizePolicy(QSizePolicy::Minimum,
```

Comment:
What is line 62 doing?

Reply:
Set the size policy of the box in the OverlayPanel. If its size policy isn't QSizePolicy::Minimum, the OverlayPanel can't expand to fit its content (The text of the list item is elided) when the width of the list item is wider than the minimum width of the DropDownList.

QSizePolicy::Minimum means the sizeHint() is minimal, and sufficient. The widget cannot be smaller than the size provided by sizeHint().

The box in the OverlayPanel contains the DropDownList which includes the ScrollableListBox, setting QSizePolicy::Minimum to the box can make the width of OverlayPanel is not shorter than the width hint of ScrollableListBox which is equal to the width hint of the list item.

Reply:
My concern is that this breaks encapsulation. Anytime you have one class poking into the implementation details of another class, no aspect of that class can ever be relied on in the future and it starts a path down brittle software.

Consider some alternative ways of solving this that don't require breaking encapsulation.